### PR TITLE
dialects: (riscv) add `func` to `riscv_func` lowering

### PR DESCRIPTION
**Pull request by adutilleul**
Just a very barebone lowering that will be used by the toy pipeline.

### CODE REVIEW

File: tests/filecheck/backend/riscv/riscv_func_to_riscv_func.mlir [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+        func.return
+    }
+}
+
+// CHECK:       builtin.module {
+// CHECK-NEXT:      "riscv_func.func"() ({
+// CHECK-NEXT:          "test.op"() : () -> ()
+// CHECK-NEXT:          "riscv_func.return"() : () -> ()
+// CHECK-NEXT:      }) {"sym_name" = "main"} : () -> ()
+// CHECK-NEXT:  }
```
Comment (superlopuh) [inline tests/filecheck/backend/riscv/riscv_func_to_riscv_func.mlir:15]:
```suggestion
// CHECK-NEXT:  }

```

**Comment by superlopuh** [comment]
Seems ready for review

File: xdsl/backend/riscv/lowering/lower_func_riscv_func.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+        if name != "main":
+            raise NotImplementedError("Only support lowering main function for now")
+
+        if op.body.blocks[0].args:
+            raise NotImplementedError(
+                "Only support functions with no arguments for now"
+            )
+
+        new_func = riscv_func.FuncOp(
+            op.sym_name.data, rewriter.move_region_contents_to_new_regions(op.body)
```
Comment (AntonLydike) [inline xdsl/backend/riscv/lowering/lower_func_riscv_func.py:27]: I did not know we had this API, TIL! Very nice!

File: xdsl/backend/riscv/lowering/lower_func_riscv_func.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+        rewriter.replace_matched_op(riscv_func.ReturnOp(()))
+
+
+class LowerFuncToRiscvFunc(ModulePass):
+    name = "lower-func-to-riscv-func"
+
+    def apply(self, ctx: MLContext, op: ModuleOp) -> None:
+        PatternRewriteWalker(LowerFuncOp()).rewrite_module(op)
+        PatternRewriteWalker(LowerFuncCallOp()).rewrite_module(op)
+        PatternRewriteWalker(LowerReturnOp()).rewrite_module(op)
```
Comment (AntonLydike) [inline xdsl/backend/riscv/lowering/lower_func_riscv_func.py:54]: Can we put these into one walker?

File: xdsl/backend/riscv/lowering/lower_func_riscv_func.py [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+            op.sym_name.data, rewriter.move_region_contents_to_new_regions(op.body)
+        )
+
+        rewriter.replace_matched_op(new_func)
+
+
+class LowerFuncCallOp(RewritePattern):
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, op: func.Call, rewriter: PatternRewriter) -> None:
+        raise NotImplementedError("Function call lowering not implemented yet")
```
Comment (AntonLydike) [inline xdsl/backend/riscv/lowering/lower_func_riscv_func.py:36]: Not sure why this is even in the PR, it just does absolutely nothing?

File: xdsl/backend/riscv/lowering/lower_func_riscv_func.py [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+
+
+class LowerReturnOp(RewritePattern):
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, op: func.Return, rewriter: PatternRewriter):
+        if op.arguments:
+            raise NotImplementedError("Only support return with no arguments for now")
+
+        rewriter.replace_matched_op(riscv_func.ReturnOp(()))
```
Comment (AntonLydike) [inline xdsl/backend/riscv/lowering/lower_func_riscv_func.py:45]: And this is here but not tested, and also quite useless imo?

File: xdsl/backend/riscv/lowering/lower_func_riscv_func.py (hunk 4 above)
Comment (superlopuh) [reply xdsl/backend/riscv/lowering/lower_func_riscv_func.py:36]: I like it, it's a nice stub for us to replace later, and then the failing unit tests are replaced with working unit tests, so we get a little win in the future also

**Review by superlopuh — APPROVED** [review]

**Pull request merged**